File: Decision Tree Regression/decision_tree.py
#import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

#import data
df = pd.read_csv('C:\\Users\\saranya.ravichandran\\Desktop\\venky\\Tech\\DS\\ML-A-Z\\Machine Learning A-Z\\Part 2 - Regression\\Section 8 - Decision Tree Regression\\Position_Salaries.csv')

#treat missing data --> no missing data here
#treat categorical data --> no categorical data here

X = df.iloc[:,1:2].values
y = df.iloc[:,2].values

#feature scaling

#decision tree regression
from sklearn.tree import DecisionTreeRegressor
regressor = DecisionTreeRegressor(random_state=0)
regressor.fit(X,y)


#predict new data

y_pred = regressor.predict(6.5)

#plot values
X_grid = np.arange(min(X),max(X),0.01)
X_grid = X_grid.reshape((len(X_grid),1))
plt.scatter(X,y,color = 'red')
plt.plot(X_grid,regressor.predict(X_grid),color = 'blue')

plt.title('Decision tree regression')
plt.xlabel('Position')
plt.ylabel('Salary')
plt.show()

File: Decision Tree Regression/decision_tree.R
#import data
df = read.csv('Position_Salaries.csv')

#treat missing data --> not needed here
#treat categorical data --> not needed here
df = df[2:3]

#Decision tree regression
#install.packages('rpart')
library(rpart)
regressor = rpart(formula = Salary ~., data = df, control = rpart.control(minsplit = 1))

#predict new data
y_pred = predict(regressor, newdata = data.frame(Level = 6.5))
#plot values
#Decision tree is a non-continuous model. So feature scaling cannot affect the model performance in any way.
#For non-continuous models, visualization is done by splitting into small grids

x_grid = seq(min(df$Level),max(df$Level),0.01)
library(ggplot2)
ggplot()+
  geom_point(aes(x=df$Level,y=df$Salary),colour='red')+
  geom_line(aes(x=x_grid,y=predict(regressor,newdata = data.frame(Level = x_grid))))+
  ggtitle('Decision tree regression')+
  xlab('position')+
  ylab('Salary')

#random forest is a team of many decision trees which can be used for accurate results
